src/NHibernate/Linq/LinqExtensionMethods.cs

```diff
@@ -77,51 +81,81 @@ public static IQueryable<T> Timeout<T>(this IQueryable<T> query, int timeout)
+		/// <param name="source">An <see cref="T:System.Linq.IQueryable`1" /> to convert to a future query.</param>
+		/// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+		/// <returns>A <see cref="IEnumerable{T}"/>.</returns>
+		/// <exception cref="T:System.ArgumentNullException"><paramref name="source" /> is <see langword="null"/>.</exception>
+		/// <exception cref="T:System.NotSupportedException"><paramref name="source" /> <see cref="IQueryable.Provider"/> is not a <see cref="INhQueryProvider"/>.</exception>
+		public static IEnumerable<TSource> ToFuture<TSource>(this IQueryable<TSource> source)
 		{
-			var nhQueryable = query as QueryableBase<T>;
-			if (nhQueryable == null)
-				throw new NotSupportedException("Query needs to be of type QueryableBase<T>");
-
-			var provider = (INhQueryProvider) nhQueryable.Provider;
-			var future = provider.ExecuteFuture(nhQueryable.Expression);
-			return (IEnumerable<T>) future;
+			if (source == null)
+			{
+				throw new ArgumentNullException(nameof(source));
+			}
+			if (!(source.Provider is INhQueryProvider provider))
+			{
+				throw new NotSupportedException($"Source {nameof(source.Provider)} must be a {nameof(INhQueryProvider)}");
+			}
+			return provider.ExecuteFuture<TSource>(source.Expression);
 		}
 
-		public static IFutureValue<T> ToFutureValue<T>(this IQueryable<T> query)
+		/// <summary>
+		/// Wraps the query in a deferred <see cref="IFutureValue{T}"/> which will trigger a batch of all pending future queries
+		/// when its <see cref="IFutureValue{T}.Value"/> is read.
+		/// </summary>
+		/// <param name="source">An <see cref="T:System.Linq.IQueryable`1" /> to convert to a future query.</param>
+		/// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+		/// <returns>A <see cref="IFutureValue{T}"/>.</returns>
+		/// <exception cref="T:System.ArgumentNullException"><paramref name="source" /> is <see langword="null"/>.</exception>
+		/// <exception cref="T:System.NotSupportedException"><paramref name="source" /> <see cref="IQueryable.Provider"/> is not a <see cref="INhQueryProvider"/>.</exception>
+		public static IFutureValue<TSource> ToFutureValue<TSource>(this IQueryable<TSource> source)
 		{
-			var nhQueryable = query as QueryableBase<T>;
-			if (nhQueryable == null)
-				throw new NotSupportedException("Query needs to be of type QueryableBase<T>");
+			if (source == null)
+			{
+				throw new ArgumentNullException(nameof(source));
+			}
+			if (!(source.Provider is INhQueryProvider provider))
+			{
+				throw new NotSupportedException($"Source {nameof(source.Provider)} must be a {nameof(INhQueryProvider)}");
+			}
+			var future = provider.ExecuteFuture<TSource>(source.Expression);
+			return new FutureValue<TSource>(() => future);
+		}
 
-			var provider = (INhQueryProvider) nhQueryable.Provider;
-			var future = provider.ExecuteFuture(nhQueryable.Expression);
-			if (future is IEnumerable<T>)
```

Comment:
This `if` was non-sens: result is always `IEnumerable<T>` since the expression is extracted from an `IQueryable<T>` without applying any aggregation on it. The implicit `else` part was indeed dead code.